// graphql-to-mongodb/lib/src/mongoDbUpdateValidation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var graphql_1 = require("graphql");
var common_1 = require("./common");
var graphQLUpdateType_1 = require("./graphQLUpdateType");
var ShouldAssert;
(function (ShouldAssert) {
    ShouldAssert[ShouldAssert["DefaultTrueRoot"] = 0] = "DefaultTrueRoot";
    ShouldAssert[ShouldAssert["True"] = 1] = "True";
    ShouldAssert[ShouldAssert["False"] = 2] = "False";
})(ShouldAssert = exports.ShouldAssert || (exports.ShouldAssert = {}));
function validateUpdateArgs(updateArgs, graphQLType, overwrite) {
    var errors = [];
    errors = errors.concat(validateNonNullableFieldsOuter(updateArgs, overwrite, graphQLType));
    if (errors.length > 0) {
        throw new graphql_1.GraphQLError(errors.join("\n"));
    }
}
exports.validateUpdateArgs = validateUpdateArgs;
function validateNonNullableFieldsOuter(updateArgs, overwrite, graphQLType) {
    var shouldAssert = !!updateArgs.setOnInsert
        ? ShouldAssert.True
        : overwrite
            ? ShouldAssert.DefaultTrueRoot
            : ShouldAssert.False;
    return validateNonNullableFields(Object.keys(updateArgs).map(function (_) { return updateArgs[_]; }), graphQLType, shouldAssert);
}
function validateNonNullableFields(objects, graphQLType, shouldAssert, path) {
    if (path === void 0) { path = []; }
    var typeFields = graphQLType.getFields();
    var errors = shouldAssert === ShouldAssert.True ? validateNonNullableFieldsAssert(objects, typeFields, path) : [];
    var overwrite = objects.map(function (_) { return _[graphQLUpdateType_1.OVERWRITE]; }).filter(function (_) { return _; })[0];
    shouldAssert = getShouldAssert(shouldAssert, overwrite);
    return errors.concat(validateNonNullableFieldsTraverse(objects, typeFields, shouldAssert, path));
}
exports.validateNonNullableFields = validateNonNullableFields;
function validateNonNullableFieldsAssert(objects, typeFields, path) {
    if (path === void 0) { path = []; }
    return Object
        .keys(typeFields)
        .map(function (key) { return ({ key: key, type: typeFields[key].type }); })
        .filter(function (field) { return common_1.isNonNullType(field.type); })
        .reduce(function (agg, field) {
        var fieldPath = path.concat([field.key]).join(".");
        var fieldValues = objects.map(function (_) { return _[field.key]; }).filter(function (_) { return _ !== undefined; });
        if (field.type instanceof graphql_1.GraphQLNonNull) {
            if (fieldValues.some(function (_) { return _ === null; }))
                return agg.concat(["Non-nullable field \"" + fieldPath + "\" is set to null"]);
            if (fieldValues.length === 0)
                return agg.concat(["Missing non-nullable field \"" + fieldPath + "\""]);
        }
        if (common_1.isListType(field.type) && !validateNonNullListField(fieldValues, field.type)) {
            return agg.concat(["Non-nullable element of array \"" + fieldPath + "\" is set to null"]);
        }
        return agg;
    }, []);
}
exports.validateNonNullableFieldsAssert = validateNonNullableFieldsAssert;
function validateNonNullListField(fieldValues, type) {
    if (type instanceof graphql_1.GraphQLNonNull) {
        if (fieldValues.some(function (_) { return _ === null; })) {
            return false;
        }
        return validateNonNullListField(fieldValues, type.ofType);
    }
    if (type instanceof graphql_1.GraphQLList) {
        return validateNonNullListField(common_1.flatten(fieldValues.filter(function (_) { return _; })), type.ofType);
    }
    return true;
}
exports.validateNonNullListField = validateNonNullListField;
function getShouldAssert(current, input) {
    if (current === ShouldAssert.True) {
        return ShouldAssert.True;
    }
    if (typeof input !== "undefined") {
        return input ? ShouldAssert.True : ShouldAssert.False;
    }
    if (current === ShouldAssert.DefaultTrueRoot) {
        return ShouldAssert.True;
    }
    return current;
}
exports.getShouldAssert = getShouldAssert;
function validateNonNullableFieldsTraverse(objects, typeFields, shouldAssert, path) {
    if (path === void 0) { path = []; }
    var keys = Array.from(new Set(common_1.flatten(objects.map(function (_) { return Object.keys(_); }))));
    return keys.reduce(function (agg, key) {
        var field = typeFields[key];
        var type = field.type;
        var innerType = common_1.getInnerType(type);
        if (!(innerType instanceof graphql_1.GraphQLObjectType) || field.resolve) {
            return agg;
        }
        var newPath = path.concat([key]);
        var values = objects.map(function (_) { return _[key]; }).filter(function (_) { return _; });
        if (common_1.isListType(type)) {
            return agg.concat(common_1.flatten(flattenListField(values, type).map(function (_) { return validateNonNullableFields([_], innerType, ShouldAssert.True, newPath); })));
        }
        else {
            return agg.concat(validateNonNullableFields(values, innerType, shouldAssert, newPath));
        }
    }, []);
}
exports.validateNonNullableFieldsTraverse = validateNonNullableFieldsTraverse;
function flattenListField(objects, type) {
    if (type instanceof graphql_1.GraphQLNonNull) {
        return flattenListField(objects, type.ofType);
    }
    if (type instanceof graphql_1.GraphQLList) {
        return flattenListField(common_1.flatten(objects).filter(function (_) { return _; }), type.ofType);
    }
    return objects;
}
exports.flattenListField = flattenListField;
